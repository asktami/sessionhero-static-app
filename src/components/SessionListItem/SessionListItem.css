.SessionListItem {
  background: none;
  display: flex;
  flex-direction: column;
  border: 0;
  font-size: smaller;
  text-decoration: none;
  box-shadow: 0px 11px 21px -8px rgba(0, 0, 0, 0.5);
  margin-bottom: 40px;
  margin-right: 20px;
  border-radius: 10px;
}

.SessionListItem:hover {
  cursor: pointer;
}

.SessionListItem,
.SessionListItem:visited {
  color: #555;
}

.SessionListItem__image {
  height: 250px;
  width: 275px;
  border-radius: 10px 10px 0 0;
  background-size: cover;
  background-position: 50% 50%;
}

.SessionListItem__details {
  padding: 10px;
  display: flex;
  flex: 1;
  flex-direction: column;
  justify-content: space-between;
  width: 275px;
}

.SessionListItem__details h2 {
  margin: 0;
}

.SessionListItem__details .SessionStarRating {
  padding: 5px 0;
  text-align: center;
  opacity: 0.85;
}

.SessionListItem__details span {
  display: block;
  text-align: center;
  opacity: 0.5;
}
