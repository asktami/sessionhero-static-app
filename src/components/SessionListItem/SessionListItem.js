import React, { Component } from 'react';
import AppContext from '../../contexts/AppContext';
import SessionApiService from '../../services/session-api-service';
import { Link } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
	getColor,
	getTime,
	getDayNumber,
	getDayName
} from '../../components/Utils/Utils';
import './SessionListItem.css';

export default class SessionListItem extends Component {
	static contextType = AppContext;

	// ALWAYS use loggedIn userId
	addToSchedule = (sessionId, userId = 1) => {
		console.log('---------- add to schedule');

		// FAKE IT WITH STATE
		//  session added to schedule
		// find that sessionId in sessionList
		// create an array that is just that record from sessionList
		// add that array to the existing scheduleList

		this.context.sessionList.forEach(session => {
			if (session.id === sessionId) {
				session.userId = 1;
			}
		});

		let newScheduleList = this.context.sessionList.filter(
			session => session.userId === userId
		);

		this.context.setScheduleList(newScheduleList);

		console.log('after ADD new scheduleList = ', this.context.scheduleList);

		// FOR REAL DB ONLY
		// Promise.all([
		// 	SessionApiService.addScheduleItem(sessionId, userId),
		// 	SessionApiService.getSchedule()
		// ])
		// 	.then(results => {
		// 		const schedule = results[1];

		// 		this.context.setScheduleList(schedule);

		// 		// in postgres use joins instead
		// 		this.updateSessionList();
		// 	})
		// 	.catch(this.context.setError);
	};

	removeFromSchedule = sessionId => {
		console.log('---------- remove from schedule');

		// FAKE IT WITH STATE
		// set removed sessionList userId to blank
		// remove session from schedule  = create new scheduleList from sessionList

		this.context.sessionList.map(session => {
			if (session.id === sessionId) {
				session.userId = '';
			}
		});

		let newScheduleList = this.context.scheduleList.filter(
			session => session.id !== sessionId
		);

		this.context.setScheduleList(newScheduleList);

		console.log('after REMOVE new scheduleList = ', this.context.scheduleList);

		// FOR REAL DB ONLY
		// Promise.all([
		// 	SessionApiService.deleteScheduleItem(scheduleId),
		// 	SessionApiService.getSchedule()
		// ])
		// 	.then(results => {
		// 		const schedule = results[1];

		// 		this.context.setScheduleList(schedule);

		// 		// in postgres use joins instead
		// 		this.updateSessionList();
		// 	})
		// 	.catch(this.context.setError);
	};

	updateSessionList() {
		const { sessionList = [], scheduleList = [] } = this.context;

		// FAKE IT
		// automatically updates sessionList in context
		sessionList.forEach(session => {
			scheduleList.forEach(schedule => {
				if (schedule.sessionId === session.id) {
					session.userId = schedule.userId;
				}
			});
		});

		console.log('sessionListItem update sessionList = ', sessionList);

		// to update scheduleList in context
		let newScheduleList = sessionList.filter(session => session.userId === 1);

		console.log('sessionListItem update newScheduleList = ', newScheduleList);

		this.context.setScheduleList(newScheduleList);
	}

	render() {
		const { setToggleId, toggleId, expandAll } = this.context;
		const { session, pathname } = this.props;

		return (
			<>
				<div className="flex-item-header-row">
					<div>
						<span className="day-name">{getDayName(session.date)}</span>{' '}
						<span className="day-of-week">{getDayNumber(session.date)}</span>
					</div>
					<div className={'track ' + getColor(session.track)}>
						{session.track}
					</div>
				</div>

				<div className="flex-row">
					<div className="flex-col-then-row cell-fixed-width-50">
						<div
							className={
								'cell-fixed-width-25 hide day-right-border border-' +
								getColor(session.track)
							}
						>
							<span className="day-name">{getDayName(session.date)}</span>&nbsp;
							<span className="day-of-week">{getDayNumber(session.date)}</span>
						</div>

						<div
							className={
								'cell-fixed-width-25 time time-right-border border-' +
								getColor(session.track)
							}
						>
							<div>
								{getTime(session.date, session.time_start)}
								<br />
								{getTime(session.date, session.time_end)}
							</div>
						</div>

						<div className="description">
							<span>
								<Link to={`/sessions/${session.id}`}>
									<span className="simple title">{session.name}</span>
								</Link>
							</span>
							<br />
							<span className="location">{session.location}</span>
						</div>
					</div>

					<div
						className={
							'flex-col-then-row hide track ' + getColor(session.track)
						}
					>
						<div className="track-text">{session.track}</div>
					</div>

					<div className="flex-col">
						{!expandAll && !pathname.includes('/sessions/') ? (
							<button
								className="btn-expand-item"
								aria-expanded="false"
								aria-label="show-session-details-button"
								onClick={() => setToggleId(session.id)}
							>
								{toggleId === session.id ? (
									<FontAwesomeIcon icon="chevron-up" size="2x" />
								) : (
									<FontAwesomeIcon icon="chevron-down" size="2x" />
								)}
							</button>
						) : null}

						{/* need to be loggedIn AND session.userId === loggedIn userId */}
						{parseInt(session.userId) === 1 ? (
							<button
								className="btn-add-to-schedule"
								aria-label="add-session-to-schedule-button"
								onClick={() => this.removeFromSchedule(session.id)}
							>
								{/* userId = {session.userId} <br /> */}
								<FontAwesomeIcon icon="star" size="2x" />
							</button>
						) : (
							<button
								className="btn-add-to-schedule"
								aria-label="add-session-to-schedule-button"
								onClick={() => this.addToSchedule(session.id)}
							>
								{/* userId = {session.userId} <br /> */}
								<FontAwesomeIcon icon={['far', 'star']} size="2x" />
							</button>
						)}
					</div>
				</div>

				<div
					className={
						'flex-footer-row toggle-content ' +
						(expandAll ||
						toggleId === session.id ||
						pathname.includes('/sessions/')
							? 'is-visible'
							: null)
					}
				>
					<div>
						<p className={'simple ' + getColor(session.track)}>
							{session.speaker}
						</p>

						<p className="simple title">Session Description</p>
						<p className="simple">{session.description}</p>

						{session.objective_1 ? (
							<>
								<p className="simple title">Session Objectives</p>

								<ul className="simple objective-list">
									{session.objective_1 ? <li>{session.objective_1}</li> : null}

									{session.objective_2 ? <li>{session.objective_2}</li> : null}

									{session.objective_3 ? <li>{session.objective_3}</li> : null}

									{session.objective_4 ? <li>{session.objective_4}</li> : null}
								</ul>
							</>
						) : null}

						{session.background ? (
							<>
								<p className="simple title">Recommended Background</p>
								<p className="simple">{session.background}</p>
							</>
						) : null}
					</div>
				</div>
			</>
		);
	}
}
