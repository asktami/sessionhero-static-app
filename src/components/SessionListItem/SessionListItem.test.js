import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import SessionListItem from './SessionListItem';

describe(`SessionListItem component`, () => {
	const props = {
		pathname: '',
		session: {
			userId: 1,
			track: 'Business',
			day: 'tue',
			date: '8/6/2019',
			time_start: '14:45:00',
			time_end: '15:15:00',
			location: 'Osceola 1 - 3',
			id: 'BUS04',
			name: 'FileMaker in Action  Custom Apps for Manufacturing and Warehouses',
			description:
				"In this session, we'll cover topics that relate well to the manufacturing/warehouse space. First, two approaches to inventory; one easy-to-implement solution with unstored calcs, and a more complex solution using static fields with scripted amount updates. Next, we'll go over the creation of labels for the warehouse and how we've used automation to save a LOT of time, including taking a 60 or 90 minute process per shipment down to 30 to 60 seconds. Optionally, as time permits in the session, we will cover some of the practicalities of using FileMaker in a warehouse space.",
			background:
				'Attendees should have a basic understanding of one to many relationships, calculation fields, and global fields.',
			objective_1: 'An easy approach to inventory using calculation fields',
			objective_2:
				'A more complex approach to inventory using script updates to static fields',
			objective_3:
				'Label creation within FileMaker to save a lot of time (looping is key)',
			objective_4: 'Practicalities of FileMaker in a warehouse space',
			speaker: 'Bradley Boggs (Chameleon Like, Inc.)'
		}
	};

	it('renders a SessionListItem by default', () => {
		const wrapper = shallow(<SessionListItem {...props} />);
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('renders the SessionListItem given props', () => {
		const wrapper = shallow(<SessionListItem {...props} />);
		expect(toJson(wrapper)).toMatchSnapshot();
	});
});
