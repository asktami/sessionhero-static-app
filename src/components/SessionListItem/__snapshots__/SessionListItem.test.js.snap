// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SessionListItem component renders a SessionListItem by default 1`] = `
<Fragment>
  <div
    className="flex-item-header-row"
  >
    <div>
      <span
        className="day-name"
      >
        Tue
      </span>
       
      <span
        className="day-of-week"
      >
        6
      </span>
    </div>
    <div
      className="track business"
    >
      Business
    </div>
  </div>
  <div
    className="flex-row"
  >
    <div
      className="flex-col-then-row cell-fixed-width-50"
    >
      <div
        className="cell-fixed-width-25 hide day-right-border border-business"
      >
        <span
          className="day-name"
        >
          Tue
        </span>
         
        <span
          className="day-of-week"
        >
          6
        </span>
      </div>
      <div
        className="cell-fixed-width-25 time time-right-border border-business"
      >
        <div>
          2:45 PM
          <br />
          3:15 PM
        </div>
      </div>
      <div
        className="description"
      >
        <span>
          <Link
            replace={false}
            to="/sessions/BUS04"
          >
            <span
              className="simple title"
            >
              FileMaker in Action  Custom Apps for Manufacturing and Warehouses
            </span>
          </Link>
        </span>
        <br />
        <span
          className="location"
        >
          Osceola 1 - 3
        </span>
      </div>
    </div>
    <div
      className="flex-col-then-row hide track business"
    >
      <div
        className="track-text"
      >
        Business
      </div>
    </div>
    <div
      className="flex-col"
    >
      <button
        aria-expanded="false"
        aria-label="show-session-details-button"
        className="btn-expand-item"
        onClick={[Function]}
      >
        <FontAwesomeIcon
          icon="chevron-down"
          size="2x"
        />
      </button>
      <button
        aria-label="add-session-to-schedule-button"
        className="btn-add-to-schedule"
        onClick={[Function]}
      >
        <FontAwesomeIcon
          icon="star"
          size="2x"
        />
      </button>
    </div>
  </div>
  <div
    className="flex-footer-row toggle-content null"
  >
    <div>
      <p
        className="simple business"
      >
        Bradley Boggs (Chameleon Like, Inc.)
      </p>
      <p
        className="simple title"
      >
        Session Description
      </p>
      <p
        className="simple"
      >
        In this session, we'll cover topics that relate well to the manufacturing/warehouse space. First, two approaches to inventory; one easy-to-implement solution with unstored calcs, and a more complex solution using static fields with scripted amount updates. Next, we'll go over the creation of labels for the warehouse and how we've used automation to save a LOT of time, including taking a 60 or 90 minute process per shipment down to 30 to 60 seconds. Optionally, as time permits in the session, we will cover some of the practicalities of using FileMaker in a warehouse space.
      </p>
      <p
        className="simple title"
      >
        Session Objectives
      </p>
      <ul
        className="simple objective-list"
      >
        <li>
          An easy approach to inventory using calculation fields
        </li>
        <li>
          A more complex approach to inventory using script updates to static fields
        </li>
        <li>
          Label creation within FileMaker to save a lot of time (looping is key)
        </li>
        <li>
          Practicalities of FileMaker in a warehouse space
        </li>
      </ul>
      <p
        className="simple title"
      >
        Recommended Background
      </p>
      <p
        className="simple"
      >
        Attendees should have a basic understanding of one to many relationships, calculation fields, and global fields.
      </p>
    </div>
  </div>
</Fragment>
`;

exports[`SessionListItem component renders the SessionListItem given props 1`] = `
<Fragment>
  <div
    className="flex-item-header-row"
  >
    <div>
      <span
        className="day-name"
      >
        Tue
      </span>
       
      <span
        className="day-of-week"
      >
        6
      </span>
    </div>
    <div
      className="track business"
    >
      Business
    </div>
  </div>
  <div
    className="flex-row"
  >
    <div
      className="flex-col-then-row cell-fixed-width-50"
    >
      <div
        className="cell-fixed-width-25 hide day-right-border border-business"
      >
        <span
          className="day-name"
        >
          Tue
        </span>
         
        <span
          className="day-of-week"
        >
          6
        </span>
      </div>
      <div
        className="cell-fixed-width-25 time time-right-border border-business"
      >
        <div>
          2:45 PM
          <br />
          3:15 PM
        </div>
      </div>
      <div
        className="description"
      >
        <span>
          <Link
            replace={false}
            to="/sessions/BUS04"
          >
            <span
              className="simple title"
            >
              FileMaker in Action  Custom Apps for Manufacturing and Warehouses
            </span>
          </Link>
        </span>
        <br />
        <span
          className="location"
        >
          Osceola 1 - 3
        </span>
      </div>
    </div>
    <div
      className="flex-col-then-row hide track business"
    >
      <div
        className="track-text"
      >
        Business
      </div>
    </div>
    <div
      className="flex-col"
    >
      <button
        aria-expanded="false"
        aria-label="show-session-details-button"
        className="btn-expand-item"
        onClick={[Function]}
      >
        <FontAwesomeIcon
          icon="chevron-down"
          size="2x"
        />
      </button>
      <button
        aria-label="add-session-to-schedule-button"
        className="btn-add-to-schedule"
        onClick={[Function]}
      >
        <FontAwesomeIcon
          icon="star"
          size="2x"
        />
      </button>
    </div>
  </div>
  <div
    className="flex-footer-row toggle-content null"
  >
    <div>
      <p
        className="simple business"
      >
        Bradley Boggs (Chameleon Like, Inc.)
      </p>
      <p
        className="simple title"
      >
        Session Description
      </p>
      <p
        className="simple"
      >
        In this session, we'll cover topics that relate well to the manufacturing/warehouse space. First, two approaches to inventory; one easy-to-implement solution with unstored calcs, and a more complex solution using static fields with scripted amount updates. Next, we'll go over the creation of labels for the warehouse and how we've used automation to save a LOT of time, including taking a 60 or 90 minute process per shipment down to 30 to 60 seconds. Optionally, as time permits in the session, we will cover some of the practicalities of using FileMaker in a warehouse space.
      </p>
      <p
        className="simple title"
      >
        Session Objectives
      </p>
      <ul
        className="simple objective-list"
      >
        <li>
          An easy approach to inventory using calculation fields
        </li>
        <li>
          A more complex approach to inventory using script updates to static fields
        </li>
        <li>
          Label creation within FileMaker to save a lot of time (looping is key)
        </li>
        <li>
          Practicalities of FileMaker in a warehouse space
        </li>
      </ul>
      <p
        className="simple title"
      >
        Recommended Background
      </p>
      <p
        className="simple"
      >
        Attendees should have a basic understanding of one to many relationships, calculation fields, and global fields.
      </p>
    </div>
  </div>
</Fragment>
`;
